neonvm/runner/main.go

```diff
@@ -359,16 +413,126 @@ func calcDirUsage(dirPath string) (int64, error) {
+		//     /swap.raw:10737418240B:file:512:512:gpt::;
+		//     1:2097152B:10736369663B:10734272512B::primary:;
+		//
+		// So, we'll parse it.
+		partinfoLines := strings.Split(strings.TrimSpace(partinfo), "\n")
+		if len(partinfoLines) != 3 {
+			return fmt.Errorf("partition info for swap.raw had %d lines, but we expected 3", len(partinfoLines))
+		}
+		swapLine := strings.Split(partinfoLines[2], ":")
+
+		expectedFields := 7
+		if len(swapLine) != expectedFields {
+			return fmt.Errorf(
+				"partition info for swap.raw had %d fields for swap partition, but we expected %d",
+				len(swapLine), expectedFields,
+			)
+		}
+		// start position *measured in bytes* is the second field (index = 1), and total size is the
+		// fourth field (index = 3).
+		partStart, err := strconv.Atoi(strings.TrimSuffix(swapLine[1], "B"))
+		if err != nil {
+			return fmt.Errorf("failed to parse start offset %q of swap partition: %w", swapLine[1], err)
+		}
+		partSize, err := strconv.Atoi(strings.TrimSuffix(swapLine[3], "B"))
+		if err != nil {
+			return fmt.Errorf("failed to parse size %q of swap partition: %w", swapLine[3], err)
+		}
+
+		// Create the swap. This might be quite large, so we'd like to make sure that this is a
+		// sparse file.
+		//
+		// It's also worth noting that, after running mkswap, the file is still sparse.
+		// Only a header is written, and the rest of the file remains filled with zeros, just like
+		// the contents of the partition in we created in 'swap.raw'.
+		//
+		// The header has size equal to a memory page, which is *almost always* 4096 bytes, but
+		// technically may be larger (e.g. M1 Macbooks have 16K pages).
+		// We can run 'getconf PAGE_SIZE' to find this value.
+		//
+		// Only copying the first 4KiB saves needing to write GiBs of data back into 'swap.raw',
+		// which would otherwise greatly extend NeonVM startup times.
+		if err := execFg("fallocate", "-l", fmt.Sprint(partSize), "swapfile"); err != nil {
+			return err
+		}
+		if err := execFg("mkswap", "-L", diskName, "swapfile"); err != nil {
+			return err
+		}
+		pagesizeString, err := execFgCapture("getconf", "PAGE_SIZE")
+		if err != nil {
+			return err
+		}
+		pagesizeString = strings.TrimSpace(pagesizeString)
+		pagesize, err := strconv.Atoi(pagesizeString)
+		if err != nil {
+			return fmt.Errorf("failed to parse 'getconf PAGE_SIZE' output %q as integer: %w", pagesizeString, err)
+		}
+
+		// Copy *exactly* the first pagesize of swapfile into the start of the partition in swap.raw
+		swapfile, err := os.Open("swapfile")
+		if err != nil {
+			return fmt.Errorf("failed to open swapfile: %w", err)
+		}
+		swapRaw, err := os.OpenFile("swap.raw", os.O_RDWR, 0)
+		if err != nil {
+			return fmt.Errorf("failed to open swap.raw: %w", err)
+		}
+		if _, err := swapRaw.Seek(int64(partStart), 0); err != nil {
+			return fmt.Errorf("failed to seek to %d in swap.raw: %w", partStart, err)
+		}
+
+		if _, err := io.CopyN(swapRaw, swapfile, int64(pagesize)); err != nil {
```

Comment:
This feels fragile

Reply:
How do you mean? AFAICT this is consistent with the format for linux swap — although I haven't found documentation for that directly, and instead got this from the implementation of `mkswap`.

Would it help to add some sources justifying this?

Reply:
Implementation changed